### PR TITLE
Implement readout protection for STM32F7

### PR DESCRIPTION
**Pull request by benediktibk**
Implement the readout protection functions for the STM32F7.

### CODE REVIEW

File: drivers/flash/flash_stm32_ex_op.c [hunk 1]
```diff
@@ -85,13 +88,79 @@ int flash_stm32_ex_op_sector_wp(const struct device *dev, const uintptr_t in,
...
+		if (enable) {
+			target_level = FLASH_STM32_RDP1;
+			if (permanent) {
+#if defined(CONFIG_FLASH_STM32_READOUT_PROTECTION_PERMANENT_ALLOW)
+				target_level = FLASH_STM32_RDP2;
+#else
+				LOG_ERR("Permanent readout protection (RDP "
+					"level 0 -> 2) not allowed");
+				return -ENOTSUP;
+#endif
```
Comment (de-nordic) [inline drivers/flash/flash_stm32_ex_op.c:121]: Not my call here, that is up to STM mainteiner, but wouldn't it be better to have the message as LOG_DBG?
The ERR messages are quite long and:
 1) I doubt they will be useful more than once, before tests figure out that code should not use the feature or the feature should be enabled;
 2) They will be always there, uselessly hanging there in rom.

Comment (benediktibk) [reply drivers/flash/flash_stm32_ex_op.c:121]: Looks like these were originally asserts, which were then changed into LOG_ERR:
https://github.com/zephyrproject-rtos/zephyr/commit/e7799eb55187f30c8755b63606f42dac937b2f77#diff-f7e9be53ed802bc0c668dc70d020bccd2ea03cba0217792ec471948f27efe336

Personally, I would have also chosen asserts. But I've got no real strong opinion on this, both options are fine by me.

**Comment by FRASTM** [comment]
thanks @benediktibk to add this for the stm32F7.

Recently there was a similar question for the stm32h7 serie here https://github.com/zephyrproject-rtos/zephyr/discussions/76297
Does it make sense to extend this PR to the stm32H7 serie  ?
The stm32H7 flash registers do not have  exactly the same mapping as the stm32F7. In any case that could be a next PR to come just after that one is merged.

**Review by FRASTM — COMMENTED** [review]
LGTM
did you run any test with CONFIG_FLASH_STM32_READOUT_PROTECTION set ?
(of course changing the RDP might be ... definitive, to be handled with care or simulated)

**Comment by benediktibk** [comment]
> LGTM
> did you run any test with CONFIG_FLASH_STM32_READOUT_PROTECTION set ?
> (of course changing the RDP might be ... definitive, to be handled with care or simulated)

Yes, I ran tests with changes between RDP level 0 and 1, none with 2. For obvious reasons.

**Comment by benediktibk** [comment]
> thanks @benediktibk to add this for the stm32F7.
> 
> Recently there was a similar question for the stm32h7 serie here https://github.com/zephyrproject-rtos/zephyr/discussions/76297
> Does it make sense to extend this PR to the stm32H7 serie  ?
> The stm32H7 flash registers do not have  exactly the same mapping as the stm32F7. In any case that could be a next PR to come just after that one is merged.

I'll look into it, but I won't be able to test it on hardware.

**Comment by MaazSk** [comment]
> > thanks @benediktibk to add this for the stm32F7.
> > Recently there was a similar question for the stm32h7 serie here #76297
> > Does it make sense to extend this PR to the stm32H7 serie  ?
> > The stm32H7 flash registers do not have  exactly the same mapping as the stm32F7. In any case that could be a next PR to come just after that one is merged.
> 
> I'll look into it, but I won't be able to test it on hardware.

I have the hardware and I am planning to look into this tomorrow.
@FRASTM, once this PR is merged I think I'll be able to have mine.

**Comment by FRASTM** [comment]
> > thanks @benediktibk to add this for the stm32F7.
> > Recently there was a similar question for the stm32h7 serie here #76297
> > Does it make sense to extend this PR to the stm32H7 serie  ?
> > The stm32H7 flash registers do not have  exactly the same mapping as the stm32F7. In any case that could be a next PR to come just after that one is merged.
> 
> I'll look into it, but I won't be able to test it on hardware.

Thanks for adding this serie. 
I have tested on stm32h750b_dk  board 

> > thanks @benediktibk to add this for the stm32F7.
> > Recently there was a similar question for the stm32h7 serie here #76297
> > Does it make sense to extend this PR to the stm32H7 serie  ?
> > The stm32H7 flash registers do not have  exactly the same mapping as the stm32F7. In any case that could be a next PR to come just after that one is merged.
> 
> I'll look into it, but I won't be able to test it on hardware.

Thanks for adding the stm32H7 serie.
I am afraid this is not as simple as I expected. I realize, the flash driver for stm32h7 is apart from other stm32 series. It does not include (build) the flash_stm32_ex_op.c nor  flash_stm32.c  (see drivers/flash/CMakeLists.txt)
Consequently the stm32h7_flash driver probably needs more than 
 the api function 	`.ex_op = flash_stm32_ex_op`, or `.ex_op = flash_stm32h7_ex_op` but also either re-defining the flash_stm32_ex_op.c  or a new flash_stm32h7_ex_op.c. Tis has to be studied further.
Sorry for that, maybe in the end we better have a dedicated PR  if that one s going too far in changes for stm32H7

**Comment by FRASTM** [comment]
I took your  commit  f2fe52c7d04f7d8682c2067887fc32b17f46fe4f  and applied the model to the stm32h7 :  this gives https://github.com/zephyrproject-rtos/zephyr/pull/76640

**Comment by benediktibk** [comment]
Thanks @FRASTM, I've removed the two commits for the H7 from this PR again.

File: drivers/flash/flash_stm32_ex_op.c (hunk 1 above)
Comment (erwango) [reply drivers/flash/flash_stm32_ex_op.c:121]: Agreed to move to LOG_DBG. Thanks for pointing this.

Comment (benediktibk) [reply drivers/flash/flash_stm32_ex_op.c]: done

File: drivers/flash/flash_stm32f7x.c [hunk 2]
```diff
@@ -159,6 +159,53 @@ int flash_stm32_write_range(const struct device *dev, unsigned int offset,
...
+	regs->OPTCR = (regs->OPTCR & ~mask) | value;
+	regs->OPTCR |= FLASH_OPTCR_OPTSTRT;
+
+	/* Make sure previous write is completed. */
+	barrier_dsync_fence_full();
+
+	rc = flash_stm32_wait_flash_idle(dev);
+	if (rc < 0) {
+		return rc;
+	}
```
Comment (de-nordic) [inline drivers/flash/flash_stm32f7x.c:190]: At this point you can just return RC.

Comment (benediktibk) [reply drivers/flash/flash_stm32f7x.c]: done

**Review by FRASTM — APPROVED** [review]

**Review by erwango — APPROVED** [review]

**Pull request merged**